// jovo-platform-alexa/dist/src/services/AlexaAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const https = require("https");
const ApiError_1 = require("./ApiError");
const jovo_core_1 = require("jovo-core");
const AlexaAPIResponse_1 = require("./AlexaAPIResponse");
class AlexaAPI {
    /**
     *
     * @param {ApiCallOptions} options
     * @returns {Promise<any>}
     */
    static async apiCall(options) {
        return new Promise((resolve, reject) => {
            const opt = {
                hostname: options.endpoint.substr(8),
                port: 443,
                path: options.path,
                method: options.method ? options.method : 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': 'Bearer ' + options.permissionToken,
                },
            };
            let postData;
            if (options.json) {
                postData = JSON.stringify(options.json);
                if (!opt.headers) {
                    opt.headers = {};
                }
                opt.headers['Accept'] = 'application/json';
                opt.headers['Content-Length'] = Buffer.byteLength(postData);
            }
            const req = https
                .request(opt, (res) => {
                res.setEncoding('utf8');
                let rawData = '';
                res.on('data', (chunk) => {
                    rawData += chunk;
                });
                res.on('end', () => {
                    let parsedData;
                    if (res.statusCode === 204) {
                        // no content
                        resolve(new AlexaAPIResponse_1.AlexaAPIResponse(res.statusCode, {}));
                        return;
                    }
                    try {
                        if (rawData.length > 0) {
                            parsedData = JSON.parse(rawData);
                            return resolve(new AlexaAPIResponse_1.AlexaAPIResponse(res.statusCode, parsedData));
                        }
                    }
                    catch (e) {
                        return reject(new ApiError_1.ApiError(e.message || 'Something went wrong', e.code || ApiError_1.ApiError.PARSE_ERROR));
                    }
                    resolve(new AlexaAPIResponse_1.AlexaAPIResponse(res.statusCode, {}));
                });
            })
                .on('error', (e) => {
                reject(e);
            });
            if (postData) {
                req.write(postData);
            }
            req.end();
        });
    }
    static progressiveResponse(speech, requestId, apiEndPoint, apiAccessToken) {
        return new Promise((resolve, reject) => {
            const outputSpeech = speech.toString();
            const json = {
                header: {
                    requestId,
                },
                directive: {
                    type: 'VoicePlayer.Speak',
                    speech: jovo_core_1.SpeechBuilder.toSSML(outputSpeech),
                },
            };
            const options = {
                hostname: apiEndPoint.substr(8),
                port: 443,
                path: '/v1/directives',
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': 'Bearer ' + apiAccessToken,
                },
            };
            const req = https.request(options, (res) => {
                res.setEncoding('utf8');
                res.on('data', (body) => { });
                res.on('end', () => {
                    resolve();
                });
            });
            req.on('error', (e) => {
                reject(new ApiError_1.ApiError(e.message));
            });
            req.write(JSON.stringify(json));
            req.end();
        });
    }
    /**
     * TODO
     * Proactive Events Authorization API call
     */
    static proactiveEventAuthorization(clientId, clientSecret) {
        return new Promise((resolve, reject) => {
            const body = `grant_type=client_credentials&client_id=${clientId}&client_secret=${clientSecret}&scope=alexa::proactive_events`;
            const options = {
                hostname: 'api.amazon.com',
                port: 443,
                path: '/auth/O2/token',
                method: 'POST',
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded',
                    'Content-Length': body.length,
                },
            };
            const req = https.request(options, (res) => {
                let returnData = '';
                res.setEncoding('utf8');
                res.on('data', (chunk) => {
                    returnData += chunk;
                });
                res.on('end', () => {
                    resolve(JSON.parse(returnData));
                });
            });
            req.on('error', (e) => {
                reject(new ApiError_1.ApiError(e.message));
            });
            req.write(body);
            req.end();
        });
    }
}
exports.AlexaAPI = AlexaAPI;
