// jovo-platform-googleassistant/dist/src/modules/Transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jovo_core_1 = require("jovo-core");
const _set = require("lodash.set");
const _get = require("lodash.get");
const GoogleAction_1 = require("../core/GoogleAction");
const __1 = require("..");
const GoogleActionAPI_1 = require("../services/GoogleActionAPI");
class Transaction {
    constructor(googleAction, googleAssistant) {
        this.googleAction = googleAction;
        this.googleAssistant = googleAssistant;
    }
    checkRequirements(orderOptions, paymentOptions) {
        this.googleAction.$output.GoogleAssistant = {
            TransactionRequirementsCheck: {
                orderOptions,
                paymentOptions,
            },
        };
        return this;
    }
    getRequirementsCheckResult() {
        for (const argument of _get(this.googleAction.$originalRequest || this.googleAction.$request, 'inputs[0]["arguments"]', [])) {
            if (argument.name === 'TRANSACTION_REQUIREMENTS_CHECK_RESULT') {
                return _get(argument, 'extension.resultType');
            }
        }
    }
    isRequirementsCheckOk() {
        return this.getRequirementsCheckResult() === 'OK';
    }
    isRequirementsCheckUserActionRequired() {
        return this.getRequirementsCheckResult() === 'USER_ACTION_REQUIRED';
    }
    isRequirementsCheckAssistantSurfaceNotSupported() {
        return this.getRequirementsCheckResult() === 'ASSISTANT_SURFACE_NOT_SUPPORTED';
    }
    isRequirementsCheckRegionNotSupported() {
        return this.getRequirementsCheckResult() === 'REGION_NOT_SUPPORTED';
    }
    askForDeliveryAddress(reason = '') {
        this.googleAction.$output.GoogleAssistant = {
            AskForDeliveryAddress: {
                reason,
            },
        };
        return this;
    }
    getDeliveryAddressDecision() {
        for (const argument of _get(this.googleAction.$originalRequest || this.googleAction.$request, 'inputs[0]["arguments"]', [])) {
            if (argument.name === 'DELIVERY_ADDRESS_VALUE') {
                return _get(argument, 'extension.userDecision');
            }
        }
    }
    isDeliveryAddressAccepted() {
        return this.getDeliveryAddressDecision() === 'ACCEPTED';
    }
    isDeliveryAddressRejected() {
        return this.getDeliveryAddressDecision() === 'REJECTED';
    }
    getDeliveryAddress() {
        for (const argument of _get(this.googleAction.$originalRequest || this.googleAction.$request, 'inputs[0]["arguments"]', [])) {
            if (argument.name === 'TRANSACTION_REQUIREMENTS_CHECK_RESULT') {
                return _get(argument, 'extension.resultType');
            }
        }
    }
    getDeliveryAddressLocation() {
        if (!this.isDeliveryAddressAccepted()) {
            return;
        }
        for (const argument of _get(this.googleAction.$originalRequest || this.googleAction.$request, 'inputs[0]["arguments"]', [])) {
            if (argument.name === 'DELIVERY_ADDRESS_VALUE') {
                return _get(argument, 'extension.location');
            }
        }
    }
    transactionDecision(orderOptions, paymentOptions, proposedOrder) {
        this.googleAction.$output.GoogleAssistant = {
            TransactionDecision: {
                orderOptions,
                paymentOptions,
                proposedOrder,
            },
        };
        return this;
    }
    createOrder(speech, orderUpdate) {
        this.googleAction.$output.GoogleAssistant = {
            OrderUpdate: {
                orderUpdate,
                speech,
            },
        };
    }
    getTransactionDecisionResult() {
        for (const argument of _get(this.googleAction.$originalRequest || this.googleAction.$request, 'inputs[0]["arguments"]', [])) {
            if (argument.name === 'TRANSACTION_DECISION_VALUE') {
                return _get(argument, 'extension.userDecision');
            }
        }
    }
    isOrderAccepted() {
        return this.getTransactionDecisionResult() === 'ORDER_ACCEPTED';
    }
    isOrderRejected() {
        return this.getTransactionDecisionResult() === 'ORDER_REJECTED';
    }
    isDeliveryAddressUpdated() {
        return this.getTransactionDecisionResult() === 'DELIVERY_ADDRESS_UPDATED';
    }
    isCartChangeRequested() {
        return this.getTransactionDecisionResult() === 'CART_CHANGE_REQUESTED';
    }
    getSubscriptions(skus) {
        return this.getSkus(skus, 'SKU_TYPE_SUBSCRIPTION');
    }
    getConsumables(skus) {
        return this.getSkus(skus, 'SKU_TYPE_IN_APP');
    }
    completePurchase(skuId) {
        this.googleAction.$output.GoogleAssistant = {
            CompletePurchase: {
                skuId,
            },
        };
    }
    getPurchaseStatus() {
        for (const argument of _get(this.googleAction.$originalRequest || this.googleAction.$request, 'inputs[0]["arguments"]', [])) {
            if (argument.name === 'COMPLETE_PURCHASE_VALUE') {
                return argument.extension.purchaseStatus;
            }
        }
    }
    getSkus(skus, type) {
        const conversationId = _get(this.googleAction.$request, 'originalDetectIntentRequest.payload.conversation.conversationId');
        if (!this.googleAssistant.config.transactions ||
            !this.googleAssistant.config.transactions.androidPackageName) {
            throw new jovo_core_1.JovoError('getSkus needs the Android App package name', jovo_core_1.ErrorCode.ERR, 'jovo-platform-googleassistant');
        }
        const promise = this.getGoogleApiAccessToken().then((accessToken) => {
            return GoogleActionAPI_1.GoogleActionAPI.apiCall({
                endpoint: 'https://actions.googleapis.com',
                path: `/v3/packages/${this.googleAssistant.config.transactions.androidPackageName}/skus:batchGet`,
                method: 'POST',
                permissionToken: accessToken,
                json: {
                    conversationId,
                    skuType: type,
                    ids: skus,
                },
            });
        });
        return promise.then((response) => response.data);
    }
    getGoogleApiAccessToken() {
        if (!this.googleAssistant.config.transactions ||
            !this.googleAssistant.config.transactions.keyFile) {
            throw new jovo_core_1.JovoError('Please add a valid keyFile object to the GoogleAssistant transaction config', jovo_core_1.ErrorCode.ERR, 'jovo-platform-googleassistant');
        }
        try {
            const googleapis = require('googleapis').google;
            return googleapis.auth
                .getClient({
                keyFile: this.googleAssistant.config.transactions.keyFile,
                scopes: ['https://www.googleapis.com/auth/actions.purchases.digital'],
            })
                .then((client) => client.authorize())
                .then((authorization) => authorization.access_token);
        }
        catch (e) {
            console.log(e);
            console.log(e.stack);
            if (e.message === "Cannot find module 'googleapis'") {
                return Promise.reject(new jovo_core_1.JovoError(e.message, jovo_core_1.ErrorCode.ERR, 'jovo-platform-googleassistant', undefined, 'Please run `npm install googleapis`'));
            }
            return Promise.reject(new Error('Could not retrieve Google API access token'));
        }
    }
}
exports.Transaction = Transaction;
class TransactionsPlugin {
    install(googleAssistant) {
        googleAssistant.middleware('$type').use(this.type.bind(this));
        googleAssistant.middleware('$output').use(this.output.bind(this));
        this.googleAssistant = googleAssistant;
        GoogleAction_1.GoogleAction.prototype.$transaction = undefined;
    }
    type(googleAction) {
        if (_get(googleAction.$originalRequest || googleAction.$request, 'inputs[0].intent') ===
            'actions.intent.TRANSACTION_REQUIREMENTS_CHECK') {
            _set(googleAction.$type, 'type', 'ON_TRANSACTION');
            _set(googleAction.$type, 'subType', 'TRANSACTION_REQUIREMENTS_CHECK');
        }
        if (_get(googleAction.$originalRequest || googleAction.$request, 'inputs[0].intent') ===
            'actions.intent.DELIVERY_ADDRESS') {
            _set(googleAction.$type, 'type', 'ON_TRANSACTION');
            _set(googleAction.$type, 'subType', 'DELIVERY_ADDRESS');
        }
        if (_get(googleAction.$originalRequest || googleAction.$request, 'inputs[0].intent') ===
            'actions.intent.TRANSACTION_DECISION') {
            _set(googleAction.$type, 'type', 'ON_TRANSACTION');
            _set(googleAction.$type, 'subType', 'TRANSACTION_DECISION');
        }
        if (_get(googleAction.$originalRequest || googleAction.$request, 'inputs[0].intent') ===
            'actions.intent.TRANSACTION_DECISION') {
            _set(googleAction.$type, 'type', 'ON_TRANSACTION');
            _set(googleAction.$type, 'subType', 'COMPLETE_PURCHASE');
        }
        googleAction.$transaction = new Transaction(googleAction, this.googleAssistant);
    }
    output(googleAction) {
        if (!googleAction.$originalResponse) {
            googleAction.$originalResponse = new __1.GoogleActionResponse();
        }
        const output = googleAction.$output;
        if (_get(output, 'GoogleAssistant.TransactionRequirementsCheck')) {
            _set(googleAction.$originalResponse, 'expectUserResponse', true);
            _set(googleAction.$originalResponse, 'systemIntent', {
                intent: 'actions.intent.TRANSACTION_REQUIREMENTS_CHECK',
                data: {
                    '@type': 'type.googleapis.com/google.actions.v2.TransactionRequirementsCheckSpec',
                    'paymentOptions': _get(output, 'GoogleAssistant.TransactionRequirementsCheck.paymentOptions'),
                },
            });
        }
        if (_get(output, 'GoogleAssistant.AskForDeliveryAddress')) {
            _set(googleAction.$originalResponse, 'expectUserResponse', true);
            _set(googleAction.$originalResponse, 'systemIntent', {
                intent: 'actions.intent.DELIVERY_ADDRESS',
                data: {
                    '@type': 'type.googleapis.com/google.actions.v2.DeliveryAddressValueSpec',
                    'addressOptions': {
                        reason: _get(output, 'GoogleAssistant.AskForDeliveryAddress.reason'),
                    },
                },
            });
        }
        if (_get(output, 'GoogleAssistant.TransactionDecision')) {
            _set(googleAction.$originalResponse, 'expectUserResponse', true);
            _set(googleAction.$originalResponse, 'systemIntent', {
                intent: 'actions.intent.TRANSACTION_DECISION',
                data: {
                    '@type': 'type.googleapis.com/google.actions.v2.TransactionDecisionValueSpec',
                    'orderOptions': {
                        requestDeliveryAddress: true,
                    },
                    'paymentOptions': _get(output, 'GoogleAssistant.TransactionDecision.paymentOptions'),
                    'proposedOrder': _get(output, 'GoogleAssistant.TransactionDecision.proposedOrder'),
                },
            });
        }
        if (_get(output, 'GoogleAssistant.OrderUpdate')) {
            _set(googleAction.$originalResponse, 'expectUserResponse', true);
            const richResponseItems = _get(googleAction.$originalResponse, 'richResponse.items', []);
            richResponseItems.push({
                structuredResponse: {
                    orderUpdate: _get(output, 'GoogleAssistant.OrderUpdate.orderUpdate'),
                },
            });
            richResponseItems.push({
                simpleResponse: {
                    textToSpeech: _get(output, 'GoogleAssistant.OrderUpdate.speech'),
                },
            });
            _set(googleAction.$originalResponse, 'richResponse.items', richResponseItems);
        }
        const completePurchase = _get(output, 'GoogleAssistant.CompletePurchase');
        if (completePurchase) {
            _set(googleAction.$originalResponse, 'expectUserResponse', true);
            _set(googleAction.$originalResponse, 'systemIntent', {
                intent: 'actions.intent.COMPLETE_PURCHASE',
                inputValueData: {
                    '@type': 'type.googleapis.com/google.actions.transactions.v3.CompletePurchaseValueSpec',
                    'skuId': completePurchase.skuId,
                },
            });
            _set(googleAction.$originalResponse, 'inputPrompt', {
                initialPrompts: [
                    {
                        textToSpeech: 'PLACEHOLDER',
                    },
                ],
                noInputPrompts: [],
            });
        }
    }
    uninstall(googleAssistant) { }
}
exports.TransactionsPlugin = TransactionsPlugin;
